File: 3519.count-numbers-with-non-decreasing-digits.py
#
# @lc app=leetcode id=3519 lang=python3
#
# [3519] Count Numbers with Non-Decreasing Digits
#

# @lc code=start
class Solution:
    def countNumbers(self, l: str, r: str, b: int) -> int:
        MOD = 10**9 + 7

        def to_base_b(num: int, base: int) -> list[int]:
            if num == 0:
                return []
            digits = []
            while num > 0:
                digits.append(num % base)
                num //= base
            digits.reverse()
            return digits

        def count_up_to(D: list[int], base: int) -> int:
            L = len(D)
            memo = {}

            def dfs(pos: int, prev: int, tight: int, started: int) -> int:
                if pos == L:
                    return 1 if started else 0
                key = (pos, prev, tight, started)
                if key in memo:
                    return memo[key]
                ans = 0
                up = D[pos] if tight else base - 1
                for d in range(up + 1):
                    new_tight = 1 if tight and d == up else 0
                    if started == 0:
                        if d == 0:
                            new_started = 0
                            new_prev = 0
                            ans = (ans + dfs(pos + 1, new_prev, new_tight, new_started)) % MOD
                        else:
                            new_started = 1
                            new_prev = d
                            ans = (ans + dfs(pos + 1, new_prev, new_tight, new_started)) % MOD
                    else:
                        if d >= prev:
                            new_started = 1
                            new_prev = d
                            ans = (ans + dfs(pos + 1, new_prev, new_tight, new_started)) % MOD
                memo[key] = ans
                return ans

            return dfs(0, 0, 1, 0)

        r_num = int(r)
        right_count = count_up_to(to_base_b(r_num, b), b)
        l_num = int(l) - 1
        left_count = count_up_to(to_base_b(l_num, b), b)
        return (right_count - left_count + MOD) % MOD

# @lc code=end

File: 3519.count-numbers-with-non-decreasing-digits.cpp
#
# @lc app=leetcode id=3519 lang=cpp
#
# [3519] Count Numbers with Non-Decreasing Digits
#
# @lc code=start
class Solution {
public:
    int countNumbers(string l, string r, int b) {
        const int MOD = 1e9 + 7;
        
        int countR = countUpTo(r, b, MOD);
        string l_minus_1 = subtract1(l);
        int countL = (l_minus_1.empty() || l_minus_1 == "0") ? 0 : countUpTo(l_minus_1, b, MOD);
        
        int result = (countR - countL + MOD) % MOD;
        return result;
    }
    
private:
    string subtract1(string s) {
        int n = s.length();
        for (int i = n - 1; i >= 0; i--) {
            if (s[i] > '0') {
                s[i]--;
                break;
            } else {
                s[i] = '9';
            }
        }
        int i = 0;
        while (i < s.length() && s[i] == '0') i++;
        return i == s.length() ? "0" : s.substr(i);
    }
    
    vector<int> toBase(string s, int b) {
        vector<int> result;
        while (s != "0" && !s.empty()) {
            int remainder = 0;
            string next = "";
            for (char c : s) {
                int digit = c - '0';
                int curr = remainder * 10 + digit;
                if (!next.empty() || curr / b > 0) {
                    next += (char)('0' + curr / b);
                }
                remainder = curr % b;
            }
            result.push_back(remainder);
            s = next.empty() ? "0" : next;
        }
        reverse(result.begin(), result.end());
        return result;
    }
    
    int countUpTo(string s, int b, int MOD) {
        vector<int> digits = toBase(s, b);
        int n = digits.size();
        
        map<tuple<int, int, int, int>, int> memo;
        
        function<int(int, int, bool, bool)> dp = [&](int pos, int last_digit, bool tight, bool started) -> int {
            if (pos == n) {
                return started ? 1 : 0;
            }
            
            auto key = make_tuple(pos, last_digit, tight, started);
            if (memo.count(key)) {
                return memo[key];
            }
            
            int limit = tight ? digits[pos] : (b - 1);
            long long result = 0;
            
            for (int digit = 0; digit <= limit; digit++) {
                bool new_tight = tight && (digit == limit);
                
                if (!started) {
                    if (digit == 0) {
                        result = (result + dp(pos + 1, 0, new_tight, false)) % MOD;
                    } else {
                        result = (result + dp(pos + 1, digit, new_tight, true)) % MOD;
                    }
                } else {
                    if (digit >= last_digit) {
                        result = (result + dp(pos + 1, digit, new_tight, true)) % MOD;
                    }
                }
            }
            
            memo[key] = result;
            return result;
        };
        
        return dp(0, 0, true, false);
    }
};
# @lc code=end